### PR TITLE
[USH-3855] N-per-row attribute for Groups

### PR DESCRIPTION
**Pull request by akashkj**
## Product Description
When `n-per-row` property is applied to a Group in webapps, it's behavior was not reflected in the UI.

image

With collapsible groups

image

## Technical Summary
https://dimagi-dev.atlassian.net/browse/USH-3855

Add a class to sub-group template of type `col-sm-X` where X is determined using the value of 'n' in n-per-row. It uses the bootstrap's grid system contaning 12 columns. Addiitonally, wrap the sub-group div in another div with class `row` to property align the components.

## Feature Flag
NA

### Safety story
Local Testing

### Automated test coverage
There are some existing tests that are failing with this change. Need to fix those.

### QA Plan
https://dimagi-dev.atlassian.net/browse/QA-5949

### Migrations
NA

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js [hunk 1]
```diff
@@ -171,7 +171,8 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
         var self = this;
 
         if (!json.type) {
-            Container.groupQuestions(json);
+            Container.groupElements(json, constants.QUESTION_TYPE);
```
Comment (Jtang-1) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:174]: Does this first group questions and then goes through again and groups the groups? 

What if a row could fit 2 groups and a question? I think all children needs to be grouped sequentially but i'm not sure. For example: if the form is like this
Question (width 4)
Question (width 4)
Group (width 4)
They should all fit on the same row. But if we group Question first, then it'll put the two questions on one row (it won't fit the group since it's still width =12).

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js [hunk 2]
```diff
@@ -312,22 +315,52 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
...
+            } else if (child.type === constants.REPEAT_TYPE) {
+                groupElementsAndResetGroup(child);
+            } else {
+                pushChildAndResetGroup(child);
+            }
+        }
+
+        for (let child of json.children) {
+            if (elementType === constants.QUESTION_TYPE) {
+                processChildForGroupingQuestions(child, usedWidth, currentGroup, newChildren);
```
Comment (Jtang-1) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js]: Is there a way we can reduce the number of sub-functions needed? I had a hard time following in particular the small functions like `groupElementsAndResetGroup` and `pushChildAndResetGroup`.

 I don't think we need two separate processes. I'm imagining this can done as a DFS with the base cases handling the different `elementTypes`/ if there isn't a `n-per-row` style and recursively calling if needed. I haven't fully thought this out so open to if you don't think it's practical.

**Review by Jtang-1 — COMMENTED** [review]
Haven't finished reviewing, will continue tomorrow. The screenshot in the description looks like what we're going for so you're on the right track.

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js (hunk 1 above)
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:174]: > Does this first group questions and then goes through again and groups the groups?

Yes

> What if a row could fit 2 groups and a question? 

That's a good question. I didn't think about this scenario and was more inclined towards the case shared in the ticket(two or more groups in a row). Right now sequential questions are grouped under 'grouped-question' and I have added logic for the groups to be under 'grouped-groups'. When there are different types of elements in the same row, there need to be a common grouping type in order to keep them in that row using `<div class='row'>`. How could that be achieved, any suggestion?

**Comment by akashkj** [comment]
@Jtang-1 I have updated the code to use a single call to group the elements.

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js [hunk 3]
```diff
@@ -314,52 +311,25 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
...
-                pushChildAndResetGroup(child);
-            }
-        }
-
-        function processChildForGroupingGroups(child) {
-            if (child.type === constants.GROUP_TYPE) {
-                addToGroupForAllowedWidth(child);
+        for (let child of json.children) {
+            if (child.type === constants.QUESTION_TYPE || child.type === constants.GROUP_TYPE) {
+                const elementTileWidth = Question.calculateColumnWidthForPerRowStyle(child.style);
```
Comment (Jtang-1) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:316]: `calculateColumnWidthForPerRowStyle` should be defined somewhere else since it's no longer specific to `Question`. Maybe it can be a part of `GroupedElementTileRow` and renamed to something like `calculateElementWidth`

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js [hunk 4]
```diff
@@ -708,8 +721,8 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
             });
```
Comment (Jtang-1) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:721]: For the function above `childrenRequired`, does it still propagate correctly now that there could be additional levels of group nesting?

The purpose of this is really to find if any question contained in this `GroupedElementTileRow` is a required question

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js [hunk 5]
```diff
@@ -635,7 +638,7 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
 
         self.hasAnyNestedQuestions = function () {
             return _.any(self.children(), function (d) {
-                if (d.type() === constants.QUESTION_TYPE || d.type() === constants.REPEAT_TYPE || d.type() === constants.GROUPED_QUESTION_TILE_ROW_TYPE) {
+                if (d.type() === constants.QUESTION_TYPE || d.type() === constants.REPEAT_TYPE || d.type() === constants.GROUPED_ELEMENT_TILE_ROW_TYPE) {
```
Comment (Jtang-1) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:641]: You'll need to check nested elements for `GROUPED_ELEMENT_TILE_ROW_TYPE` since it can contain groups

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js [hunk 6]
```diff
@@ -655,12 +658,22 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
             }
             return Container.prototype.headerBackgroundColor.call(self);
         };
+
+        const columnWidth = Question.calculateColumnWidthForPerRowStyle(this.style);
```
Comment (Jtang-1) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:662]: I think this can be deleted

**Review by Jtang-1 — COMMENTED** [review]
Thanks for the changes, this logic looks simpler to understand. Could you also add a test for this?

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js (hunk 6 above)
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:662]: It is used here - https://github.com/dimagi/commcare-hq/pull/33834/files/577da9006b958dfcbb92474a3a68b99ce0e139dd#diff-fa78ad0c884f5acb33ba17fb2142ba438de13b9bb043c65b1687111cfca48190R667

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js (hunk 3 above)
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js]: https://github.com/dimagi/commcare-hq/pull/33834/commits/50eec9c7bdb4113e5657009f9f4adc11d5c058c3

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js (hunk 4 above)
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:721]: https://github.com/dimagi/commcare-hq/pull/33834/commits/606fae2a2ee67bcdfef90adc3d5d9d126b334120

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js (hunk 5 above)
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js]: https://github.com/dimagi/commcare-hq/pull/33834/commits/b5037b04ca9a2bafe53850bf8b53622f1e18de20

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js (hunk 6 above)
Comment (Jtang-1) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js]: Oh right, i didn't notice that. Could this block become just this? And remove checks of the style. `calculateElementWidth` should return 12 (for the whole column width) if no style is defined.
```
let columnWidth = GroupedElementTileRow.calculateElementWidth(this.style);
this.questionTileWidth = `col-sm-${columnWidth}`;
 ```
 
 also a nit: can we rename the `questionTileWidth` variable name for this since it's for the group. Maybe we can rename it to `elementTile`? calling it `groupTileWidth` also seems confusing.

**Comment by Jtang-1** [comment]
I also remembered, it would be worth checking that things look okay with collapsible groups https://confluence.dimagi.com/display/commcarepublic/Advanced+Web+Apps+Formatting#AdvancedWebAppsFormatting-CollapsibleGroups

This is an example of how collapsible groups look in a form
 https://github.com/dimagi/commcare-hq/pull/33835

**Comment by nospame** [comment]
It seems like the changes here interact with https://github.com/dimagi/commcare-hq/pull/33844. Is it possible to make sure this change still allows that appearance attribute `stripe-repeats` to work? I can help with targeting for those css styles if helpful/needed.

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/utils_spec.js [hunk 7]
```diff
@@ -45,7 +45,7 @@ hqDefine("cloudcare/js/form_entry/spec/utils_spec", function () {
             [groupInRepeat] = repeat.children();
             [textInRepeat] = groupInRepeat.children();
 
-            assert.equal(groupInRepeat.caption(), null);
+            assert.equal(groupInRepeat.caption, null);
```
[the remark this answers is not in the document]
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/utils_spec.js:48]: I'm not sure why this was affected this way. I could not see its connection with the change as this is related to repeat, which is not altered.

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js [hunk 8]
```diff
@@ -638,9 +638,9 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
 
         self.hasAnyNestedQuestions = function () {
```
[the remark this answers is not in the document]
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:639]: You are right, `Question_Type` and `Group_Type` will now always be preceeded with a `Grouped_Element_Tile_Row_Type`.

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/form_ui_spec.js [hunk 9]
```diff
@@ -265,8 +265,8 @@ hqDefine("cloudcare/js/form_entry/spec/form_ui_spec", function () {
 
         it('Should find nested questions', function () {
             var form = formUI.Form(nestedGroupJSON);
-            assert.isTrue(form.children()[0].hasAnyNestedQuestions());
-            assert.isFalse(form.children()[1].hasAnyNestedQuestions());
+            assert.isTrue(form.children()[0].children()[0].hasAnyNestedQuestions());
+            assert.isFalse(form.children()[1].children()[0].children()[0].children()[0].hasAnyNestedQuestions());
```
[the remark this answers is not in the document]
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/form_ui_spec.js:324]: Corrected. Thanks.

**Comment by akashkj** [comment]
> it would be worth checking that things look okay with collapsible groups

Tested and things look fine. Have added the relevant screenshot in PR description.

**Comment by akashkj** [comment]
> It seems like the changes here interact with https://github.com/dimagi/commcare-hq/pull/33844. Is it possible to make sure this change still allows that appearance attribute stripe-repeats to work? I can help with targeting for those css styles if useful/needed.

Hi @nospame. Could you please tell steps to test this feature?

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/form_ui_spec.js [hunk 10]
```diff
@@ -142,12 +143,11 @@ hqDefine("cloudcare/js/form_entry/spec/form_ui_spec", function () {
             formJSON.tree = [g0];
             let form = formUI.Form(formJSON);
 
-            assert.equal(form.children()[0].headerBackgroundColor(), '#002f71'); //[g0]
-            assert.equal(form.children()[0].children()[0].headerBackgroundColor(), ''); //[g0-0]
-            assert.equal(form.children()[0].children()[0].children()[2].headerBackgroundColor(), '#003e96'); //[g1]
-            assert.equal(form.children()[0].children()[0].children()[2].children()[0].headerBackgroundColor(), '#004EBC'); //[g1-0]
-            assert.equal(form.children()[0].children()[0].children()[2].children()[0].children()[2].headerBackgroundColor(), '#002f71'); //[r1]
-            assert.equal(form.children()[0].children()[0].children()[2].children()[0].children()[2].children()[0].headerBackgroundColor(), ''); //[r1-0]
```
[the remark this answers is not in the document]
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/form_ui_spec.js:150]: I could not find which element is it asserting for the background color. Will check it again.

**Comment by nospame** [comment]
> > It seems like the changes here interact with #33844. Is it possible to make sure this change still allows that appearance attribute stripe-repeats to work? I can help with targeting for those css styles if useful/needed.
> 
> Hi @nospame. Could you please tell steps to test this feature?

Yep, just added them to the wiki [here](https://confluence.dimagi.com/display/commcarepublic/Advanced+Web+Apps+Formatting#AdvancedWebAppsFormatting-StripedRepeatGroups). Feel free to reach out directly if you have other questions.

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/utils_spec.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (Jtang-1) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/utils_spec.js:48]: is `groupInRepeat` pointing to the wrong element? i'm not sure. But could you track down what the cause, otherwise I think it's hiding a real bug. I don't think `groupInRepeat.caption` exists so its returning null and this assertion passes. For example, if you do `groupInRepeat.fakeprop`, it also asserts equal

[the remark this answers is not in the document]
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/utils_spec.js]: `groupInRepeat` was pointing to `grouped-element-tile-row`, so modified it to use its child, which is pointing to sub-group now that was initially created with `groupInRepeat`.

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/form_ui_spec.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/form_ui_spec.js:150]: Corrected. Due to an additional `group-element-tile-row`, statement `r1.children[0].style = styleObj;` was not applying style to the correct group.

**Comment by Jtang-1** [comment]
The collapsible group borders don't look right. They should line up to the header width and shouldn't extend to the edge of the white/grey background

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js [hunk 11]
```diff
@@ -701,20 +708,47 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
      * @param {Object} json - The JSON returned from touchforms to represent a Form
      * @param {Object} parent - The object's parent. Either a Form, Group, or Repeat.
      */
-    function GroupedQuestionTileRow(json, parent) {
+    function GroupedElementTileRow(json, parent) {
         var self = this;
         self.parent = parent;
         Container.call(self, json);
 
+        self.isRepetition = parent instanceof Repeat;
```
Comment (Jtang-1) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:716]: i don't think you need this. But instead, [this](https://github.com/dimagi/commcare-hq/blob/96d92204f04a265cecabbc34880c33e70ac79a90/corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js#L581C17-L581C17) should be changed to `parent.parent instanceof Repeat;`

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/utils_spec.js [hunk 12]
```diff
@@ -35,7 +35,7 @@ hqDefine("cloudcare/js/form_entry/spec/utils_spec", function () {
                 textInGroup = fixtures.textJSON({ix: "1,0"}),
                 group = fixtures.groupJSON({ix: "1", children: [textInGroup]}),
                 textInRepeat = fixtures.textJSON({ix: "2_0,0"}),
-                groupInRepeat = fixtures.groupJSON({ix: "2_0", children: [textInRepeat]}),
+                groupInRepeat = fixtures.groupJSON({ix: "2_0", children: [textInRepeat], caption: null}),
```
Comment (Jtang-1) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/utils_spec.js:38]: adding caption value here seems to circumvent the point of the test. That should be removed, and the change i suggested in the [other comment ](https://github.com/dimagi/commcare-hq/blob/96d92204f04a265cecabbc34880c33e70ac79a90/corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js#L581C17-L581C17) should resolve this.

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js [hunk 13]
```diff
@@ -660,12 +663,16 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
...
+        let columnWidth = GroupedElementTileRow.calculateElementWidth(this.style);
+        this.elementTile = `col-sm-${columnWidth}`;
     }
+
     Group.prototype = Object.create(Container.prototype);
     Group.prototype.constructor = Container;
 
     /**
-     * Represents a repeat group. A repeat only has Group objects as children. Each child Group contains GroupedQuestionTileRow
+     * Represents a repeat group. A repeat only has Group objects as children. Each child Group contains GroupedElementTileRow
```
Comment (Jtang-1) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:675]: This comment doesn't seem accurate. Could you also update the other comments to reflect this new hierarchy where all `Group` are contained within a `GroupElementTileRow`

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/utils_spec.js (hunk 12 above)
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/utils_spec.js]: https://github.com/dimagi/commcare-hq/pull/33834/commits/dbab40c2c0c15313bc355da9910ac8ee96dcb24d

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js (hunk 11 above)
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js]: https://github.com/dimagi/commcare-hq/pull/33834/commits/dbab40c2c0c15313bc355da9910ac8ee96dcb24d

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js (hunk 13 above)
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js]: https://github.com/dimagi/commcare-hq/pull/33834/commits/17693864e29fabcaeb75052c839361d3299f9194

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js [hunk 14]
```diff
@@ -576,7 +580,7 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
     Form.prototype.constructor = Container;
 
     /**
-     * Represents a group of GroupedQuestionTileRow which contains questions.
+     * Represents a group of GroupedElementTileRow which contains questions.
```
Comment (Jtang-1) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:583]: nit: I think it should be "which contains Questions or Groups". The different uses of the word "group" is getting confusing, maybe using capital "Group" to represent the object would help.

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/utils_spec.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (Jtang-1) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/utils_spec.js]: resolved here https://github.com/dimagi/commcare-hq/pull/33834#discussion_r1430650776

**Review by Jtang-1 — COMMENTED** [review]
Thanks for your work on this Akash, it's looking good. 
Only additional comment is resolving the styles of the group that QA found https://dimagi-dev.atlassian.net/browse/QA-5973 and could you also check if the missing x-icon is related to this?

I'm not sure when this went through QA, but if it was before the substantial changes were made, it would be safer to ask for another sanity check.

File: corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less [hunk 15]
```diff
@@ -67,7 +67,7 @@
   }
 
   .stripe-repeats {
-    .gr {
+    > .row, .panel-body > .children > .row {
       &:nth-of-type(odd) {
```
Comment (Jtang-1) [inline corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less:71]: should this target the `.gr` element so the resulting target is the same as before? So like 
```
 &:nth-of-type(odd) {
    .gr {
         background-color: @table-bg-accent;
      }
  }
  ```

Comment (akashkj) [reply corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less:71]: @nospame fya

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js (hunk 14 above)
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js]: https://github.com/dimagi/commcare-hq/pull/33834/commits/22c152e7d701ff9c3c4aa94c0a58a62154771463

**Comment by akashkj** [comment]
> resolving the styles of the group that QA found https://dimagi-dev.atlassian.net/browse/QA-5973 and could you also check if the missing x-icon is related to this?

Added change for collapsible group spacing https://github.com/dimagi/commcare-hq/pull/33834/commits/26253d8be5cb9f05c3a170745cd114834bab7680

> I'm not sure when this went through QA, but if it was before the substantial changes were made, it would be safer to ask for another sanity check.

There are no major changes I think since the last QA testing, but will still ask for a quick sanity check.

File: corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less [hunk 16]
```diff
@@ -208,3 +210,8 @@
     margin-bottom: 0px !important;
   }
 }
+
+div[class*="gr"][class*="col-sm-"][class*="panel"][class*="panel-default"] {
```
Comment (Jtang-1) [inline corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less:214]: I'm not familiar with this, what is this targeting? For `gr`, `panel`, and `panel-default`, is there a need to be searching as part of a substring?

Comment (akashkj) [reply corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less:214]: It needs to search for substring `col-sm-` while others are exact strings. I'm not sure if there is a way to concatenate classes with exact names and a sub-string in a common styling. Therefore, used substring for all.

Updated the image in description.

Comment (Jtang-1) [reply corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less:214]: Is the image in the description up to date now? I still see the problem with the collapsible group borders in the description image. I also don't think this targets the right thing. It shouldn't need to target any elements that have only a `col-sm...` class

Comment (akashkj) [reply corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less:214]: It's updated now in the description. Somehow it was still showing the old image.

> It shouldn't need to target any elements that have only a col-sm... class

It is targeting the `div` that contains the header and body of the group. Since `col-sm-` is quite generic, attributes are applied to combination of all 4 classes in that `div`. Do you have any alternate suggestion?

image

Comment (Jtang-1) [reply corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less:214]: Would this work?
```
.question-tile-row {
  .gr {
    padding-right: 0px !important;
    padding-left: 0px !important;
  }
}
```

Comment (akashkj) [reply corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less]: https://github.com/dimagi/commcare-hq/pull/33834/commits/173c955a6bb7cfa5bc3a1cafe54bdca211a32719

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js [hunk 17]
```diff
@@ -726,6 +726,8 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
         self.parent = parent;
         Container.call(self, json);
 
+        self.rel_ix = ko.observable(relativeIndex(self.ix()));
```
Comment (Jtang-1) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:729]: why was this needed?

File: corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less (hunk 15 above)
Comment (Jtang-1) [reply corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less:71]: is there a resolution to this? My guess is we want to target the same elements as before. I think the way it is done now extends past the border that we want it to. I'm not sure if what i suggested does it correctly, can you verify?

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js (hunk 17 above)
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:730]: There was console error on clicking 'X' for a repeat group - `rel_ix is not a function`

File: corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less (hunk 15 above)
Comment (akashkj) [reply corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less:77]: With current change, it looks like:
Screenshot from 2024-01-18 14-57-32

With suggested change, it looks like:
Screenshot from 2024-01-18 14-58-58

In this, there seem to be a border cut-off from left and right.

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js (hunk 17 above)
Comment (Jtang-1) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:730]: Was the error related to this [part](https://github.com/dimagi/commcare-hq/blob/60f2cfb1ca875b6383ca8f3f4f0a881cfdcc2924/corehq/apps/cloudcare/static/cloudcare/js/form_entry/web_form_session.js#L440)? I'm guessing you don't need this anymore after implementing this [commit](https://github.com/dimagi/commcare-hq/pull/33834/commits/80730cfcd989e15bf033d808ffbc08a5056c599f), can you give it a try?

What is the output of `relativeIndex` and what is it used for? I'm wondering mostly because I don't think it currently supports the `ix` values of `GroupedElementTileRow`. The `ix` value would look something like `(0,0_0,1)` and `relativeIndex` currently doesn't handle the paranthesis.

File: corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less (hunk 15 above)
Comment (Jtang-1) [reply corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less:77]: I see, hm i think currently it would also be cut-off on the left and right. I think the full width does look better. @nospame do you have an opinion? I see in the screenshot of the original [PR](https://github.com/dimagi/commcare-hq/pull/33844) that there's also a left and right gap. Wanted to check it wasn't for a particular reason.

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js (hunk 17 above)
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js]: Yes, it works now without having this statement. Removed it.

**Comment by akashkj** [comment]
QAT is completed today

**Review by Jtang-1 — APPROVED** [review]
Thanks for going back and forth with me on this and implementing the changes on the PR. It's looking good

File: corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less (hunk 15 above)
Comment (nospame) [reply corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/form.less:77]: @Jtang-1 my bad on not keeping up on Github notifications. No, there's no particular reason for the left and right gap, that was just how it worked out with the elements I was able to easily/logically target.
